### PR TITLE
Adding in natural language queries extension

### PR DESCRIPTION
**Pull request by BennuFire**

### CODE REVIEW

**Comment by nielsdejong** [comment]
Add filter for OpenAPI key storage in state

File: src/extensions/ExtensionsModal.tsx
Comment (AleSim94) [inline src/extensions/ExtensionsModal.tsx:1]: dunno what is happening here?

File: src/extensions/query-translator/QueryTranslatorConfig.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+interface QueryTranslatorConfig {
+  availableClients: AvailableClients;
+}
+
+export const QUERY_TRANSLATOR_CONFIG: QueryTranslatorConfig = {
+  availableClients: {
+    OpenAI: {
+      clientName: 'OpenAI',
+      clientClass: OpenAiClient,
```
Comment (AleSim94) [inline src/extensions/query-translator/QueryTranslatorConfig.ts:41]: Maybe be consistent and rename the class to OpenAIClient?

File: src/extensions/query-translator/README.md
Comment (AleSim94) [inline src/extensions/query-translator/README.md:1]: should this text be updated?

File: src/extensions/query-translator/clients/ModelClient.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        throw Error(`The model couldn't translate your request: ${inputMessage}`);
+      }
+    } catch (error) {
+      await consoleLogAsync('Error during query', error);
+      throw error;
+    }
+    return [query, newHistory];
+  }
+
+  async validateQuery(_message, _database) {
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/ModelClient.ts:179]: some things are not implemented?

File: src/extensions/query-translator/component/OverrideCardQueryEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+import { getLastMessage } from '../state/QueryTranslatorSelector';
+import { GPT_LOADING_ICON } from './LoadingIcon';
+import {
+  deleteSessionStoragePrepopulationReportFunction,
+  setSessionStoragePrepopulationReportFunction,
+} from '../../state/ExtensionActions';
+import { getPrepopulateReportExtension } from '../../state/ExtensionSelectors';
+import { LanguageIconOutline } from '@neo4j-ndl/react/icons';
+
+// TODO - rename to 'Node Sidebar Extension button' to reflect better the functionality.
```
Comment (AleSim94) [inline src/extensions/query-translator/component/OverrideCardQueryEditor.tsx:22]: part of 2.3.0? the renaming?

File: src/extensions/query-translator/component/QueryTranslator.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { SideNavigationItem } from '@neo4j-ndl/react';
+import QueryTranslatorSettingsModal from './QueryTranslatorSettingsModal';
+import { Tooltip } from '@mui/material';
+import { LanguageIconSolid } from '@neo4j-ndl/react/icons';
+/**
+ * //TODO:
```
Comment (AleSim94) [inline src/extensions/query-translator/component/QueryTranslator.tsx:8]: Still a TODO?

File: src/extensions/query-translator/component/QueryTranslatorSettingsModal.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  modelProvider,
+  clientSettings,
+  updateClientSettings,
+  updateModelProvider,
+  initializeModelClient,
+}) => {
+  const [modelProviderState, setModelProviderState] = React.useState(modelProvider);
+  const [settingsState, setSettingsState] = React.useState(clientSettings);
+
+  // TODO: a user shouldn't be able to save a configuration if it's not correct and it didn't fill all the requirements
```
Comment (AleSim94) [inline src/extensions/query-translator/component/QueryTranslatorSettingsModal.tsx:24]: Still a TODO for 2.3.0?

File: src/extensions/query-translator/state/QueryTranslatorReducer.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/**
+ * Reducers define changes to the application state when a given action
+ */
+
+import {
+  SET_MODEL_PROVIDER,
+  SET_GLOBAL_MODEL_CLIENT,
```
Comment (AleSim94) [inline src/extensions/query-translator/state/QueryTranslatorReducer.ts:7]: not used

File: src/extensions/query-translator/util/Util.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { createNotification } from '../../../application/ApplicationActions';
+import { updateReportQueryThunk } from '../../../card/CardThunks';
+import { queryTranslationThunk } from '../state/QueryTranslatorThunks';
+
+/**
+ * TODO: placeholder of function that gets injected before report populating logic.
```
Comment (AleSim94) [inline src/extensions/query-translator/util/Util.ts:6]: still a TODO for 2.3?

File: src/extensions/query-translator/util/Util.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ */
+export function translateQuery(driver, dispatch, pagenumber, id, type, extensions, setResult) {
+  // TODO - only trigger the translation if the latest english wasn't already translated, or if the english query is ''.
```
Comment (AleSim94) [inline src/extensions/query-translator/util/Util.ts:9]: still a TODO for 2.3?

File: src/extensions/query-translator/README.md
Comment (nielsdejong) [reply src/extensions/query-translator/README.md:1]: Agree, or perhaps deleted, as it's now in the docs

File: src/extensions/query-translator/clients/OpenAiClient.ts [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  constructor(settings) {
+    super(settings);
+  }
+
+  async validateQuery(message, database) {
+    let query = message.content;
+    let isValid = false;
+    let errorMessage = '';
+    try {
+      let res = await this.queryDatabase(`EXPLAIN ${query}`, database);
```
Comment (nielsdejong) [inline src/extensions/query-translator/clients/OpenAiClient.ts:28]: would there be any way to 'escape' the explain and do some kind of query injection here?
Just theorizing if there's a possiblity of edits being done to the DB here

File: src/extensions/query-translator/clients/OpenAiClient.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+  setListAvailableModels(listModels) {
+    this.listAvailableModels = listModels;
+  }
+
+  setModelType(modelType) {
+    this.modelType = modelType;
+  }
+
+  // TODO: adapt to the new structure, no more persisting inside the object, passign everything down
+  /* addUserMessage(content, reportType, plain = false) {
```
Comment (nielsdejong) [inline src/extensions/query-translator/clients/OpenAiClient.ts:110]: should we remove these comments?

File: src/extensions/query-translator/clients/const.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
+export const nodePropsQuery = `CALL apoc.meta.data()
+YIELD label, other, elementType, type, property
+WHERE NOT type = "RELATIONSHIP" AND elementType = "node"
+WITH label AS nodeLabels, collect(property) AS properties
+RETURN {labels: nodeLabels, properties: properties} AS output
+`;
+
+export const relPropsQuery = `
+CALL apoc.meta.data()
```
Comment (nielsdejong) [inline src/extensions/query-translator/clients/const.ts:9]: we should add in the extension description (in ExtensionModal) that APOC is a prereq. for using the extension. it's theoretically possible people try the extension without APOC core.

File: src/extensions/query-translator/clients/const.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+Treemap - Two variables named Path(a list of strings) and value(a numerical value).
+Radar Chart - Multiple variables representing property values of nodes and relationships.
+Sankey Chart - Three variables, two being a node object (and not a property value) and one representing a relationship object (and not a property value).
+Map - multiple variables representing nodes objects(should contain spatial propeties) and relationship objects.
+Single Value - A single value of a single variable.
+Gauge Chart - A single value of a single variable.
+Raw JSON - The Cypher query must return a JSON object that will be displayed as raw JSON data.
+Pie Chart - Two variables named category and value.`;
+
+export const MAX_NUM_VALIDATION = 3;
```
Comment (nielsdejong) [inline src/extensions/query-translator/clients/const.ts:104]: we should also document this somewhere in the app itself.
perhaps in the config screen for the extension.
So the user knows we make between 1 and 3 calls per translation request.

File: src/extensions/query-translator/component/LoadingIcon.tsx [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+export const GPT_LOADING_ICON = (
+  <div className='centered' style={{ textAlign: 'center' }}>
+    <br />
+    <img
+      style={{ width: 40, animation: 'pulse 2s infinite', marginTop: 'auto', marginLeft: 'auto', marginRight: 'auto' }}
+      src='https://seeklogo.com/images/O/open-ai-logo-8B9BFEDC26-seeklogo.com.png'
```
Comment (nielsdejong) [inline src/extensions/query-translator/component/LoadingIcon.tsx:8]: probably we should store the logo locally inside `public/openai.png`.

File: src/extensions/query-translator/component/OverrideCardQueryEditor.tsx (hunk 3 above)
Comment (nielsdejong) [reply src/extensions/query-translator/component/OverrideCardQueryEditor.tsx:22]: I'd say this is for 2.3.1 or later

**Review by nielsdejong — CHANGES_REQUESTED** [review]

File: src/extensions/query-translator/QueryTranslatorConfig.ts (hunk 1 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/QueryTranslatorConfig.ts:41]: The name that you see there is the now that will be shown to the final user, so I prefer sticking to just OpenAi

File: src/extensions/query-translator/clients/OpenAiClient.ts (hunk 10 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/clients/OpenAiClient.ts:110]: removed

File: src/extensions/query-translator/clients/const.ts (hunk 11 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/clients/const.ts:9]: Where we want to add it?

File: src/extensions/query-translator/util/Util.ts (hunk 7 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/util/Util.ts:6]: removed

File: src/extensions/query-translator/state/QueryTranslatorReducer.ts (hunk 6 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/state/QueryTranslatorReducer.ts:7]: removed

File: src/extensions/query-translator/component/LoadingIcon.tsx (hunk 13 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/component/LoadingIcon.tsx:8]: done

File: src/dashboard/DashboardThunks.ts [hunk 14]
```diff
@@ -323,6 +323,8 @@ export function upgradeDashboardVersion(dashboard: any, origin: string, target:
       styling: {
         active: true,
       },
+      active: true,
```
[the remark this answers is not in the document]
Comment (alfredorubin96) [reply src/dashboard/DashboardThunks.ts:326]: still used, maybe to remove next iteration

File: src/extensions/query-translator/clients/OpenAiClient.ts (hunk 9 above)
Comment (alfredorubin96) [reply src/extensions/query-translator/clients/OpenAiClient.ts:28]: I really don't know, but how the can comment out the EXPLAIN ?

**Pull request merged**